rust-runtime/aws-smithy-http-client/src/test_util/wire.rs

```diff
@@ -199,60 +194,115 @@ pub struct WireMockServer {
+        if let Some(graceful) = graceful {
+            graceful.shutdown().await;
+        }
+    }
+}
+
 impl WireMockServer {
     /// Start a wire mock server with the given events to replay.
     pub async fn start(mut response_events: Vec<ReplayedEvent>) -> Self {
-        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
-        let (tx, rx) = oneshot::channel();
+        let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
+        let (tx, mut rx) = oneshot::channel();
         let listener_addr = listener.local_addr().unwrap();
         response_events.reverse();
         let response_events = Arc::new(Mutex::new(response_events));
         let handler_events = response_events;
         let wire_events = Arc::new(Mutex::new(vec![]));
         let wire_log_for_service = wire_events.clone();
         let poisoned_conns: Arc<Mutex<HashSet<SocketAddr>>> = Default::default();
-        let make_service = make_service_fn(move |connection: &AddrStream| {
+        let graceful = SharedGraceful::new();
+        let conn_builder = Arc::new(hyper_util::server::conn::auto::Builder::new(
+            TokioExecutor::new(),
+        ));
+
+        let server = async move {
             let poisoned_conns = poisoned_conns.clone();
             let events = handler_events.clone();
             let wire_log = wire_log_for_service.clone();
-            let remote_addr = connection.remote_addr();
-            tracing::info!("established connection: {:?}", connection);
-            wire_log.lock().unwrap().push(RecordedEvent::NewConnection);
-            async move {
-                Ok::<_, Infallible>(service_fn(move |_: http_02x::Request<hyper_0_14::Body>| {
-                    if poisoned_conns.lock().unwrap().contains(&remote_addr) {
-                        tracing::error!("poisoned connection {:?} was reused!", &remote_addr);
-                        panic!("poisoned connection was reused!");
-                    }
-                    let next_event = events.clone().lock().unwrap().pop();
-                    let wire_log = wire_log.clone();
-                    let poisoned_conns = poisoned_conns.clone();
-                    async move {
-                        let next_event = next_event
-                            .unwrap_or_else(|| panic!("no more events! Log: {:?}", wire_log));
-                        wire_log
-                            .lock()
-                            .unwrap()
-                            .push(RecordedEvent::Response(next_event.clone()));
-                        if next_event == ReplayedEvent::Timeout {
-                            tracing::info!("{} is poisoned", remote_addr);
-                            poisoned_conns.lock().unwrap().insert(remote_addr);
-                        }
-                        tracing::debug!("replying with {:?}", next_event);
-                        let event = generate_response_event(next_event).await;
-                        dbg!(event)
+            loop {
```

Comment:
Just curious, what caused you to work harder (i.e. the code in the `loop` block)? Is it to support graceful shutdown in hyper1.x?

Reply:
Yeah the hyper 1.x landscape changed for servers as well and utility functions like `Sender` and `Graceful` were either removed or relocated. I had to spend a bit of time figuring out how to use the new APIs while retaining the previous behaviors and not changing the code too much.

Reply:
Awesome job BTW moving test utils into the new crate.